### PR TITLE
Feat/fix model class for openai phi 1238

### PR DESCRIPTION
**Pull request by jacobweiss2305**
- refactor the larger functions
- remove assistant_message and replace with agent_message
- add more cookbook examples

### CODE REVIEW

File: cookbook/providers/openai/README.md [hunk 1]
```diff
@@ -1,15 +1,15 @@
-# OpenAI Cookbook
+# OpenAI Chat function calling
```
Comment (ashpreetbedi) [inline cookbook/providers/openai/README.md:1]: Lets keep it as `OpenAI Cookbook`

File: cookbook/providers/openai/README.md [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
-### 1. Create a virtual environment
+### 1. Create and activate a virtual environment
 
 ```shell
 python3 -m venv ~/.venvs/aienv
 source ~/.venvs/aienv/bin/activate
 ```
 
-### 2. Export `OPENAI_API_KEY`
+### 2. Export your OpenaiChat API Key
```
Comment (ashpreetbedi) [inline cookbook/providers/openai/README.md:12]: same lets keep Export `OPENAI_API_KEY`

File: cookbook/providers/openai/README.md [hunk 3]
```diff
@@ -18,11 +18,59 @@ export OPENAI_API_KEY=***
 ### 3. Install libraries
 
 ```shell
-pip install -U openai phidata
+pip install -U openai duckduckgo-search duckdb yfinance exa_py phidata
 ```
 
-### 4. Test Agent
+### 4. Run Agent
```
[the remark this answers is not in the document]
Comment (ashpreetbedi) [reply cookbook/providers/openai/README.md:24]: agree `basic_agent.py` and `basic_agent_streaming.py` are good. The problem with `agent.py` is that we might run into namespace issues

File: cookbook/providers/openai/agent.py [hunk 4]
```diff
@@ -1,10 +1,16 @@
...
 agent = Agent(
     model=OpenAIChat(model="gpt-4o"),
+    tools=[DuckDuckGo()],
+    show_tool_calls=True,
+    # debug_mode=True,
 )
 
+
+# Return the result as a string
 run: RunResponse = agent.run("Share a healthy breakfast recipe")  # type: ignore
```
[the remark this answers is not in the document]
Comment (ashpreetbedi) [reply cookbook/providers/openai/agent.py:14]: I've fixed the typing issues! Yay!!

@jacobweiss2305 agree we should use a news search that requires a function call

File: cookbook/providers/openai/data_analyst.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+duckdb_tools = DuckDbTools(create_tables=False, export_tables=False, summarize_tables=False)
+duckdb_tools.create_table_from_path(
+    path="https://phidata-public.s3.amazonaws.com/demo_data/IMDB-Movie-Data.csv", table="movies"
+)
+
+agent = Agent(
+    model=OpenAIChat(model="gpt-4o"),
+    tools=[duckdb_tools],
+    show_tool_calls=True,
+    add_to_system_prompt="""
```
Comment (ashpreetbedi) [inline cookbook/providers/openai/data_analyst.py:14]: @jacobweiss2305 @ysolanky what's a better variable for `add_to_system_prompt` -- it seems clear but a bit too crude. Looking for suggestions that looks neater like: `additional_information`, `additional_context`

I do like `additional_context` what do you think?

File: phi/model/openai/chat.py [hunk 6]
```diff
@@ -1,5 +1,5 @@
-import httpx
 from typing import Optional, List, Iterator, Dict, Any, Union, Tuple
+import httpx
```
Comment (ashpreetbedi) [inline phi/model/openai/chat.py:2]: 1 line space above httpx please

File: phi/model/openai/chat.py [hunk 7]
```diff
@@ -56,20 +64,29 @@ class OpenAIChat(Model):
...
     http_client: Optional[httpx.Client] = None
     client_params: Optional[Dict[str, Any]] = None
-    # -*- Provide the OpenAI client manually
+
+    # OpenAI clients
     client: Optional[OpenAIClient] = None
     async_client: Optional[AsyncOpenAIClient] = None
+
     # Deprecated: will be removed in v3
     openai_client: Optional[OpenAIClient] = None
```
Comment (ashpreetbedi) [inline phi/model/openai/chat.py:73]: lets remove this now, Model classes aren't used anywhere :)

File: phi/model/openai/chat.py [hunk 8]
```diff
@@ -231,75 +306,55 @@ async def ainvoke_stream(self, messages: List[Message]) -> Any:
...
+                agent_message.metrics["total_tokens"] = total_tokens
                 self.metrics["total_tokens"] = self.metrics.get("total_tokens", 0) + total_tokens
 
-        # -*- Add assistant message to messages
-        messages.append(assistant_message)
-        assistant_message.log()
-
-        # -*- Parse and run tool calls
-        if assistant_message.tool_calls is not None and self.run_tools:
+    def _handle_tool_calls(
```
Comment (ashpreetbedi) [inline phi/model/openai/chat.py:340]: i think its okay for these functions to be without a leading `_` now, wdyt? 

`def handle_tool_calls()`

File: phi/model/openai/chat.py [hunk 9]
```diff
@@ -323,261 +378,167 @@ def response(self, messages: List[Message]) -> ModelResponse:
...
+                model_response.content += f"\n{calls_str}\n\n"
 
             function_call_results = self.run_function_calls(function_calls_to_run)
             if len(function_call_results) > 0:
                 messages.extend(function_call_results)
-            # -*- Get new response using result of tool call
+            return model_response
+        return None
+
+    def _create_agent_message(
```
Comment (ashpreetbedi) [inline phi/model/openai/chat.py:390]: same, can be without the leading `_`

Also should the be `create_assistant_message` as the role here is `assistant` ?

**Review by ashpreetbedi — COMMENTED** [review]
This is an amazing PR @jacobweiss2305, left some comments. Lets hack together tomorrow on this with @ysolanky

File: phi/model/openai/chat.py (hunk 8 above)
Comment (jacobweiss2305) [reply phi/model/openai/chat.py:340]: Its a python best practice:

- The method or variable is intended for internal use within the class or module.

- It's a "weak" form of name mangling - it doesn't prevent access from outside the class, but it's a strong hint that the method should not be called directly.

File: cookbook/providers/openai/data_analyst.py (hunk 5 above)
Comment (jacobweiss2305) [reply cookbook/providers/openai/data_analyst.py:14]: +1 additional_context

File: cookbook/providers/openai/agent.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobweiss2305) [reply cookbook/providers/openai/agent.py:14]: Added please pull and test latest commit

File: cookbook/providers/openai/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacobweiss2305) [reply cookbook/providers/openai/README.md:24]: Added please pull and test latest commit

File: cookbook/providers/openai/README.md (hunk 1 above)
Comment (jacobweiss2305) [reply cookbook/providers/openai/README.md]: Done

File: cookbook/providers/openai/README.md (hunk 2 above)
Comment (jacobweiss2305) [reply cookbook/providers/openai/README.md]: Done

File: cookbook/providers/openai/data_analyst.py (hunk 5 above)
Comment (jacobweiss2305) [reply cookbook/providers/openai/data_analyst.py:14]: This should be PR for agent class. happy to make commit once we all agree

File: phi/model/openai/chat.py (hunk 6 above)
Comment (jacobweiss2305) [reply phi/model/openai/chat.py:3]: Done

File: phi/model/openai/chat.py (hunk 8 above)
Comment (ashpreetbedi) [reply phi/model/openai/chat.py:367]: good one :)

**Pull request merged**